README.md
[![Build Status](https://travis-ci.org/spark/softap-setup-js.svg)](https://travis-ci.org/spark/softap-setup-js)
[![Open Issues](https://img.shields.io/github/issues/spark/softap-setup-js.svg)](https://github.com/particle-iot/softap-setup-js/issues)

# softap-setup-js
Particle SoftAP Setup for JavaScript — Perform wireless setup of Particle devices with Node.js or a browser.

See also [softap-setup-ts, a community-created Typescript port of the SoftAP Setup library](https://github.com/markterrill/softap-setup-ts).

## Installation

```
npm install softap-setup
```

## Configuration

When instantiating a SoftAPSetup object, simply provide an options object with values for the keys you would like to override.

### Defaults

```
{
	"host": "192.168.0.1",
	"keepAlive": true,
	"timeout": 2000,
	"noDelay": true,
	"channel": 6,
	"protocol": "tcp",
	"port": 5609 (will default to 80 if protocol = http and no port specified)
}
```

## Usage

### Protocols

This library supports both TCP and HTTP protocols for performing SoftAP setup. TCP is the default protocol. If you would like to use the HTTP protocol instead, simply override the `protocol` configuration parameter.

#### Example:
```js
var SoftAPSetup = require('softap-setup');
var sap = new SoftAPSetup({ protocol: 'http' });

sap.deviceInfo(callback);
function callback(err, dat) {
	if (err) { throw err; }
	console.log("Device ID: %s, claimed: %s", dat.id, dat.claimed ? "yes" : "no");
};
```
The above code will substitute the standard TCP port 5609 for HTTP on port 80. All methods function exactly the same regardless of chosen protocol. TCP is still the recommended protocol unless browser compatibility is specifically required.

### Obtaining Device Information

You may often find it necessary to query the ID and claim status of your device. This is accomplished with a single method: `deviceInfo`. This method takes a single argument, which should be a reference to a callback function with the standard Node.js callback signature (example below).

#### Example:
```js
var SoftAPSetup = require('softap-setup');
var sap = new SoftAPSetup();

sap.deviceInfo(callback);
function callback(err, dat) {
	if (err) { throw err; }
	console.log("Device ID: %s, claimed: %s", dat.id, dat.claimed ? "yes" : "no");
};
```
The above code will attempt to retrieve the device information from the device (assuming the system running this code is connected to the device in SoftAP mode), with the default configuration (192.168.0.1:5609). When the underlying request is successfully fulfilled, the callback is called with no error parameter (null), and a result object containing the `id` of the device as a string of hex, as well as the `claimed` status as a boolean value, like so:

```json
{ "id": "STRINGOFHEXREPRESENTINGYOURDEVICEIDHERE", "claimed": false }
```

### Obtaining the Device's Public Key

The public key must be obtained from the device before it can be successfully configured. This is due to the fact that the public key is used to encrypt the passphrase for any AP that is configured to use security (see below). If you do not already know the public key of the device, you may request it from the device with the following command:

#### Example
```js
var sap = new SoftAPSetup();
sap.publicKey(callback);
function callback(err, dat) {
	if (err) { throw err; }
	console.log(dat);
};
```

### Setting the Device's Claim Code

The device must be provided with a claim code before it can be registered to a user's account. This is a string that is typically provided by a call to the [Particle API](https://docs.particle.io/reference/api/#create-a-claim-code). Once such a claim code has been obtained, it can be provided to the device like so:

#### Example
```js
var code = "somekindaclaimcode";
var sap = newSoftAPSetup();
sap.setClaimCode(code, callback);
function callback(err, dat) {
	if(err) { throw err; }
	console.log(dat);
};
```

### Scanning for Nearby Access Points

While connected to the device in SoftAP mode, it is possible to request a list of access points that have been detected in the area. This is done by way of the `scan` command. This command is one of the few commands that will typically take more than a few hundred milliseconds to complete. When executed, the device will listen for access points which are broadcasting their SSID. Important to note here is that it's not possible to detect networks which don't broadcast their SSID. You can still configure a non-broadcast network manually (see below).

#### Example

```js
var sap = new SoftAPSetup();
sap.scan(callback);
function callback(err, dat) {
	if(err) { throw err; }
	console.log("Networks Identified:");
	console.log(dat);
};
```

### Configuring the Device for a Selected Access Point

The following code snippet will store the provided details on the device, and cause it to attempt to connect to the AP you specify. If it is unsuccessful; it will return to SoftAP mode shortly thereafter, so you may reconnect to it and try again.

#### Example
```js
var sap = new SoftAPSetup();
sap.configure({
	ssid: "<NETWORK NAME>",
	security: "<SECURITY TYPE (see below)>",
	password: "[PASSWORD]",
	channel: "<CHANNEL>"
}, callback);
```

#### Example: WPA Enterprise with PEAP/MSCHAPv2 authentication
```js
var sap = new SoftAPSetup();
sap.configure({
    ssid: "<NETWORK NAME>",
    security: "<SECURITY TYPE (see below)>",
    username: "[USERNAME]",
    password: "[PASSWORD]",
    outer_identity: "[OUTER IDENTITY (optional)]",
    ca: "[CA CERTIFICATE (in PEM format, optional)]",
    channel: "<CHANNEL>"
}, callback);
```

#### Example: WPA Enterprise with EAP-TLS authentication
```js
var sap = new SoftAPSetup();
sap.configure({
    ssid: "<NETWORK NAME>",
    security: "<SECURITY TYPE (see below)>",
    client_certificate: "[CLIENT CERTIFICATE (in PEM format)]",
    private_key: "[PRIVATE KEY (in PEM format)]",
    outer_identity: "[OUTER IDENTITY (optional)]",
    ca: "[CA CERTIFICATE (in PEM format, optional)]",
    channel: "<CHANNEL>"
}, callback);
```

### Connecting to a Previously Configured Access Point

Once you have successfully issued a `configure` command, it's now only a matter of giving the device the go-ahead to actually connect. As you may have guessed, this is done via the `connect` command. It takes only a callback parameter, and will always execute "successfully". Since there is no way to verify that the provided configuration is correct until a connection attempt is made; you will need to verify that the device is able to successfully connect to the cloud (most likely via an API request to the cloud to check for the presence of the device ID that was just configured).

#### Example
```js
var sap = new SoftAPSetup();
sap.connect(callback);
function callback(err, dat) {
	console.log("Device is attempting to connect to the AP...");
};
```

### Wireless Security Types

Valid security types are as follows:

1. "open" or "none" - no security
2. "wep_psk" - WEP pre-shared key
3. "wep_shared" - Open WEP
4. "wpa_tkip" — WPA with TKIP
5. "wpa_aes" — WPA with AES
6. "wpa2_tkip" - WPA2 with TKIP
7. "wpa2_aes" — WPA2 with AES
8. "wpa2_mixed" — WPA2 AES & TKIP
9. "wpa_enterprise_aes" — WPA Enterprise with AES
10. "wpa_enterprise_tkip" — WPA Enterprise with TKIP
11. "wpa2_enterprise_aes" — WPA2 Enterprise with AES
12. "wpa2_enterprise_tkip" — WPA2 Enterprise with TKIP
13. "wpa2_enterprise_mixed" — WPA2 Enterprise with AES/TKIP
14. "wpa2_enterprise" — WPA2 Enterprise with AES/TKIP

## Notes

### AP Password Security

It's worth noting that this library uses the public key of the device to encrypt any AP passwords that are sent when configuring and connecting your device.

### EAP-TLS Private Key Security

This library also uses the public key of the device and a random AES encryption key to encrypt the Private Key when configuring your device to connect to WPA Enterprise access point with EAP-TLS authentication.

## Running in the Browser

It's possible to do SoftAP configuration from within a web browser. However, you must first convert the SoftAP code (and all of it's dependencies) from a Node.js module into a single javascript file.

### Browserify

Install browserify:
```js
npm install -g browserify
```
From the softap-setup-js code directory, run:
```js
browserify index.js -s SoftAPSetup -o softap-browser.js
```
This will create a browser-friendly ```softap-browser.js``` file that exports the ```SoftAPSetup``` object. The only difference with the browser version of this object is that it does *NOT* support reading the configuration from a file. All other methods described above will work.

**NOTE:** *Only the "http" protocol works in the browser. "tcp" will fail because the browser does not allow direct access to sockets.*

#### Example:
```js
<!doctype html>
<html lang="en">
<head></head>
<body>
  <script src="softap-browser.js"></script>
  <script> 
    var sap = new SoftAPSetup();

    sap.deviceInfo(callback);
    function callback(err, dat) {
	    if (err) { throw err; }
	    console.log("Device ID: %s, claimed: %s", dat.id, dat.claimed ? "yes" : "no");
    };
  </script>
 </body>
 </html>
```
The above code will print the device info to the javascript console of the browser.


CHANGELOG.md
# changelog

## 4.1.0 - 30 June 2017

* Support for WPA Enterprise WiFi setup
* Minimum version of node  bumped from 0.10 to 4. 

## 4.0.3 - 15 Feb 2017

* Open a dummy socket connection on first TCP command to work around
a firmware [glitch](https://github.com/particle-iot/firmware/issues/1252)

## 4.0.2 - 15 Feb 2017

* [#17](https://github.com/particle-iot/softap-setup-js/pull/17) change license from LGPL to Apache 2.0
* [#13](https://github.com/particle-iot/softap-setup-js/pull/13) default timeout increase from 2s to 8s for TCP
* [#13](https://github.com/particle-iot/softap-setup-js/pull/13) TCP response can be spread over multiple packets
* [#24](https://github.com/particle-iot/softap-setup-js/issues/24) Configuration defaults move into SoftAP class so they are available for TCP and HTTP
* [#18](https://github.com/particle-iot/softap-setup-js/issues/18) Device ID change to lowercase
* [#23](https://github.com/particle-iot/softap-setup-js/issues/23) deviceInfo request was failing with timeout


## 4.0.1 - 26 Feb 2016

* Explicitly specify 'http' as protocol

## 4.0.0 - 23 Feb 2016

* Remove environment variable and file configuration options.
* Rename `keep_alive` to `keepAlive` and `no_delay` to `noDelay`.
* Implement a strict timeout.
* Improve error handling.

## 3.0.2 - 15 Oct 2015

* Include bugfix by indraastra for CORS requests without preflights.


softap.js
module.exports = SoftAP;

var net = require('net');
var util = require('util');
var http = require('http');
var rsa = require('node-rsa');

var defaults = {
	host: "192.168.0.1",
	keep_alive: true,
	timeout: 2000,
	no_delay: true,
	default_channel: 6,
	default_protocol: "http",
	available_protocols: {
		tcp: {
			port: "5609"
		},
		http: {
			port: "80"
		}
	}
};	
var config = defaults;


var securityTable = {
	"open": 0,
	"none": 0,
	"wep_psk": 1,
	"wep_shared": 0x8001,
	"wpa_tkip": 0x00200002,
	"wpa_aes": 0x00200004,
	"wpa2_aes": 0x00400004,
	"wpa2_tkip": 0x00400002,
	"wpa2_mixed": 0x00400006,
	"wpa2": 0x00400006
};

// hashtag lazyJS
function is(cb) {

	if (cb && typeof cb == 'function') { return true }
	throw new Error('Invalid callback function provided.');
};

function SoftAP(opts) {

	if(opts && typeof opts == 'object') {
		Object.keys(opts).forEach(function _loadOpts(key) {
			config[key] = opts[key];
		});
	}

	this.protocol = config['protocol'];
	this.keepAlive = config['keep_alive'];
	this.noDelay = config['no_delay'];
	this.timeout = config['timeout'];

	this.host = config['host'];
	this.port = config['port'];

	if(!this.protocol) {
		this.protocol = config['default_protocol'];
	}
	if(!this.port) {
		this.port = defaults.available_protocols[this.protocol].port;
	}

	this.__publicKey = undefined;

	return this;
};

SoftAP.prototype.scan = function scan(cb) {

	is(cb);
	var sock = this.__sendCommand('scan-ap', cb);
	return sock;
};

SoftAP.prototype.connect = function connect(cb) {

	is(cb);
	var sock = this.__sendCommand({ name: 'connect-ap', body: { idx: 0 } }, cb);
	return sock;
};

SoftAP.prototype.deviceInfo = function deviceInfo(cb) {

	is(cb);
	var sock = this.__sendCommand('device-id', response.bind(this));
	function response(err, dat) {

		if(err) { return cb(err); }

		var claimed = dat.c === '1' ? true : false;
		this.__deviceID = dat.id;

		cb(null, {
			id : dat.id,
			claimed : claimed
		});
	};
	return sock;
};

SoftAP.prototype.publicKey = function publicKey(cb) {

	is(cb);
	var sock = this.__sendCommand('public-key', response.bind(this));
	function response(err, dat) {

		if(err) { return cb(err); }
		if(!dat) { return cb(new Error('No data received')); }
		if(dat.r !== 0) {
			return cb(new Error('Received non-zero response code'));
		}
		var buff = new Buffer(dat.b, 'hex');
		this.__publicKey = new rsa(buff.slice(22), 'pkcs1-public-der', {
			encryptionScheme: 'pkcs1'
		})
		cb(null, this.__publicKey.exportKey('pkcs8-public'));
	};
	return sock;
};

SoftAP.prototype.setClaimCode = function(code, cb) {

	is(cb);
	if(!code || typeof code !== "string") {
		throw new Error('Must provide claim code string as first parameter');
	}
	var claim = {
		k: "cc"
		, v: code
	};
	var sock = this.__sendCommand({ name: 'set', body: claim }, cb);

	return sock;
};

SoftAP.prototype.configure = function configure(opts, cb) {

	is(cb);

	var securePass = undefined;

	if(!this.__publicKey) {
		throw new Error('Must retrieve public key of device prior to AP configuration');
	}
	if(!opts || typeof opts !== 'object') {
		throw new Error('Missing configuration options object as first parameter');
	}
	if(!opts.ssid) {
		if(!opts.name) {
			throw new Error('Configuration options contain no ssid property');
		}
		opts.ssid = opts.name;
	}

	if((opts.enc || opts.sec) && !opts.security) {
		opts.security = opts.sec || opts.enc;
	}
	if(!opts.security) {
		opts.security = "open";
		opts.password = null;
	}
	if(opts.password || opts.pass) {
		if(!opts.security) {
			throw new Error('Password provided but no security type specified');
		}
		if(opts.pass && !opts.password) {
			opts.password = opts.pass;
		}
		securePass = this.__publicKey.encrypt(opts.password, 'hex');
	}
	if(typeof opts.security === "string") {
		opts.security = securityTable[opts.security];
	}

	var apConfig = {
		idx: 0,
		ssid: opts.ssid,
		sec: opts.security,
		ch: parseInt(opts.channel)
	};

	if(securePass) { apConfig.pwd = securePass; }

	var sock = this.__sendCommand({ name: 'configure-ap', body: apConfig }, cb);

	return sock;
};

SoftAP.prototype.__getSocket = function __getSocket(connect, data, error) {

	var errorMessage = undefined;
	if(typeof connect !== 'function') {
		errorMessage = "Invalid connect function specified.";
	}
	if(typeof data !== 'function') {
		errorMessage = "Invalid data function specified.";
	}
	if(error && typeof error !== 'function') {
		errorMessage = "Provided error handler is not a function.";
	}
	if(errorMessage) { throw new Error(errorMessage); }

	var sock = net.createConnection(this.port, this.host);

	sock.setTimeout(this.timeout);

	sock.on('data', data);
	if(error) { sock.on('error', error); }
	sock.on('connect', connect);

	return sock;
};

SoftAP.prototype.__httpRequest = function __httpRequest(cmd, data, error) {

	var sock;
	var payload;
	var errorMessage = undefined;

	if(!cmd || typeof cmd !== "object") {
		errorMessage = "Invalid command object specified.";
	}
	if(errorMessage) { throw new Error(errorMessage); }

	var opts = {
		method: 'GET',
		path: '/' + cmd.name,
		hostname: this.host,
		port: this.port,
		protocol: this.protocol ? this.protocol + ":" : ''
	};

	if((cmd.body) && typeof cmd.body === 'object') {
		payload = JSON.stringify(cmd.body);
		// NOTE: 'Content-Type' is set here to make this a "simple" cross-site
		// request, as per the HTTP CORS docs:
		//   https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS#Simple_requests
		// According to the spec, this means that POST can be made directly
		// without an OPTIONS request being made first.
		opts.headers = {
			'Content-Length': payload.length,
			'Content-Type': 'application/x-www-form-urlencoded'
		};
		opts.method = 'POST';
	}

	sock = http.request(opts, function responseHandler(res) {
		var results = '';
		res.on('data', function dataHandler(chunk) {
			if(chunk) { results += chunk.toString(); }
		});
		res.on('end', function () {
			data(results);
		});
	});

	sock.on('error', error);
	payload && sock.write(payload);
	sock.end();

	return sock;
};

SoftAP.prototype.__sendCommand = function(cmd, cb) {

	var sock;
	var protocol = this.protocol;

	if(typeof cmd == 'string') {
		cmd = { name : cmd, body : undefined };
	}
	else if (typeof cmd == 'object') {
		if(!cmd.name) { throw new Error('Command object has no name property'); }
	}
	else { throw new Error('Invalid command'); }
	is(cb);

	if(protocol == "http") {
		sock = this.__httpRequest(cmd, onData, cb);
	}
	else {
		sock = this.__getSocket(tcpConnected, onData);
	}

	function tcpConnected() {

		if((cmd.body) && typeof cmd.body === 'object') {

			var body = JSON.stringify(cmd.body);
			var length = body.length;
			send = util.format("%s\n%s\n\n%s", cmd.name, length, body);
		}
		else {

			send = util.format("%s\n0\n\n", cmd.name);
		}

		sock.write(send);
	};

	function onData(dat) {
		if(dat instanceof Buffer || typeof dat === 'string') {
			try {
				var json = JSON.parse(dat.toString());
			}
			catch (e) {
				return cb(new Error('Invalid JSON received from device.'));
			}
		}
		else if(typeof dat === 'object') {
			var json = dat;
		}

		cb(null, json);
	};
	return sock;
};

SoftAP.prototype.version = function(cb) {

	is(cb);
	var sock = this.__sendCommand('version', cb);
	return sock;
};

SoftAP.prototype.securityLookup = function(dec) {

	var match = null;
	Object.keys(securityTable).forEach(function(key) {
		if(parseInt(dec) == securityTable[key]) {
			match = key;
		}
	});
	return match;
};
